### PR TITLE
feat!: use byte types for bytes proto fields

### LINKED ISSUE #49: use []byte or custom type for pubsub Message data
## Expected Behavior

```go
// Option 1
type Message struct {
	Data        []byte        `json:"data,omitempty"`        // The binary data in the message.
}
```

```go
// Option 2
type Message struct {
	Data        ByteData        `json:"data,omitempty"`        // The binary data in the message.
}
```

## Actual Behavior

```go
type Message struct {
	Data        *string        `json:"data,omitempty"`        // The binary data in the message.
}
```

## Details

The choice to use a string was also made in our HTTP/REST libraries. We did not respect the [”format”: “byte”](https://github.com/googleapis/google-api-go-client/blob/7570391cc25b18bef231a82fadc38f78fa76265c/pubsub/v1/pubsub-api.json#L1779). We have an [open issue](https://github.com/googleapis/google-api-go-client/issues/346) around how it would have been nice to generate these types as []byte.

Option 2 might be nice do you could provide a custom mashaller for the type that auto encodes/decodes base64 values so the user of this library does not have to.

### PR DESCRIPTION
**Pull request by grant**
Fixes https://github.com/googleapis/google-cloudevents-go/issues/49

## Summary

This breaking change adds native `[]byte` array types to proto `byte` fields. (There are around 7 of them.) As a result, it is easier for users to decode byte data this way. 

## Features

- Uses `[]byte` type for fields where appropriate for native base64 unmarshalling.
- Updates code sample to use this updated field type.
- Add clear features documentation with list of features (such as native base64 decoding).

## Note

The post-gen script should be heavily documented, and the transformation functions (like `fixByteFields`) should be in a functional coding style (inputs and outputs).

### CODE REVIEW

File: tools/src/postgen-bytetypes.ts
Comment (jskeet) [inline tools/src/postgen-bytetypes.ts:40]: A more reliable alternative to this would be to use the compiled proto descriptor. I don't know what the Go protobuf library is like, but certainly the C# equivalent wouldn't be terribly complicated. (We use the compiled proto descriptor to extract CloudEvent types in C#, for example.)

I can imagine all kinds of places where this could go wrong, e.g. by parsing a comment as if it were a field declaration.

I'd be okay with this as a temporary measure if we believe that the schema generator + QuickType are likely to handle this properly in the future, but I wouldn't want this to be in place for the long term.

**Review by jskeet — COMMENTED** [review]
Aside: I generally find it easier to review PRs where the generation change is in one commit, and then the results are in a second commit.

This definitely looks better to me as a non-Go-user, although I have concerns around the mechanism. Definitely worth waiting for an actual Go developer to approve.

**Review by codyoss — APPROVED** [review]
LGTM from the Go perspective, thank you! I let someone else LGTM the TypeScript 😄

File: tools/src/postgen-bytetypes.ts
Comment (grant) [reply tools/src/postgen-bytetypes.ts:40]: > A more reliable alternative to this would be to use the compiled proto descriptor. I don't know what the Go protobuf library is like, but certainly the C# equivalent wouldn't be terribly complicated. (We use the compiled proto descriptor to extract CloudEvent types in C#, for example.)

I've thought about alternative solutions to achieve this Go behavior and chose this route as the best design all things considered. A long-term solution would be to change a couple things: the JSON schema to use the new (draft 7) `'contentEncoding': 'base64'`and the Golang generator to read this field and apply the correct type. However, that isn't in place today. There are a couple downstream libraries that need updates in order to support this.

We could use the `@grpc/proto-loader` for the proto part of this fix to use a compiled proto descriptor, but that wouldn't be able to modify the generated Go source code and would have a minor benefit. 

> I can imagine all kinds of places where this could go wrong, e.g. by parsing a comment as if it were a field declaration.

This isn't AST parsing; yes there are places where it could go wrong that we prevent. We could log out the ~7 places this fix modifies the Go byte fields. I have tried to make the post-generation step as readable as possible.

> I'd be okay with this as a temporary measure if we believe that the schema generator + QuickType are likely to handle this properly in the future, but I wouldn't want this to be in place for the long term.

Yes, the goal would be for better schemas and generators for base64 data. This requires a few upstream changes like https://github.com/quicktype/quicktype/issues/138.

**Comment by grant** [comment]
> Aside: I generally find it easier to review PRs where the generation change is in one commit, and then the results are in a second commit.

In general, I try to separate the concerns of commits, but while developing, there are multiple cycles of iteration, and it's good to keep track of that history.

Here's a filter for just the go files, where you can see the difference clearly:

https://github.com/googleapis/google-cloudevents-go/pull/58/files?file-filters%5B%5D=.go

File: tools/src/postgen-bytetypes.ts
Comment (grayside) [inline tools/src/postgen-bytetypes.ts:1]: I see this is already a dependency, but I thought I'd point out it's 16 lines of code with a single maintainer and no updates in 2 years.

**Review by grayside — CHANGES_REQUESTED** [review]
This doesn't look like the right approach to me. See comment about adding the proto as a dependency of code generation instead of schema generation.

**Comment by grant** [comment]
TODO:

- Add test to compile go
- Add test to unmarshal individual types

**Review by grayside — APPROVED** [review]
Discussed outstanding concerns and it seems pragmatic.

Follow-up in this PR:

* make sure the code can compile.

Follow-up before calling this library stable

* Add end-to-end tests for event types

Follow-up in the future

* Make JSON Schema the source of truth for Base64 format

File: tools/src/postgen-bytetypes.ts
Comment (grant) [reply tools/src/postgen-bytetypes.ts:53]: Yes. I've added a detailed comment as to this requirement to explain why we match on this substring.

**Comment by grant** [comment]
Thanks for the comments folks.

Per sync, created follow-up GitHub issues:

- https://github.com/googleapis/google-cloudevents-go/issues/59
- https://github.com/googleapis/google-cloudevents-go/issues/60

Go developer interface looks good. PR is approved. Merging.

**Pull request merged**
